RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc

```diff
@@ -0,0 +1,419 @@
+              << " detacalo "                 << vars[19] << std::endl;
+    if (withIso_) {
+      std::cout << " ele_pfPhotonIso "          << vars[20] << std::endl
+                << " ele_pfChargedHadIso "      << vars[21] << std::endl
+                << " ele_pfNeutralHadIso "      << vars[22] << std::endl
+                << " rho "                      << vars[23] << std::endl
+                << " preShowerOverRaw "         << vars[24] << std::endl;
+    }
+    else {
+      std::cout << " rho "                      << vars[20] << std::endl
+                << " preShowerOverRaw "         << vars[21] << std::endl;
+    }
+    std::cout << " ### MVA " << result << std::endl << std::endl;
+  }
+
+  return result;
+}
+
+int ElectronMVAEstimatorRun2Fall17::findCategory( const edm::Ptr<reco::Candidate>& particle) const {
+
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
+  if( eleRecoPtr.get() == nullptr ) {
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+  }
+  return findCategory(eleRecoPtr.get());
+}
+
+int ElectronMVAEstimatorRun2Fall17::findCategory( const reco::GsfElectron * eleRecoPtr ) const {
+  float pt = eleRecoPtr->pt();
+  float eta = eleRecoPtr->superCluster()->eta();
+
+  //
+  // Determine the category
+  //
+  int  iCategory = UNDEFINED;
+  const float ptSplit = 10;   // we have above and below 10 GeV categories
+  const float ebSplit = 0.800;// barrel is split into two regions
+  const float ebeeSplit = 1.479; // division between barrel and endcap
+
+  if (pt < ptSplit && std::abs(eta) < ebSplit) {
+    iCategory = CAT_EB1_PT5to10;
+  }
+
+  if (pt < ptSplit && std::abs(eta) >= ebSplit && std::abs(eta) < ebeeSplit) {
+    iCategory = CAT_EB2_PT5to10;
+  }
+
+  if (pt < ptSplit && std::abs(eta) >= ebeeSplit) {
+    iCategory = CAT_EE_PT5to10;
+  }
+
+  if (pt >= ptSplit && std::abs(eta) < ebSplit) {
+    iCategory = CAT_EB1_PT10plus;
+  }
+
+  if (pt >= ptSplit && std::abs(eta) >= ebSplit && std::abs(eta) < ebeeSplit) {
+    iCategory = CAT_EB2_PT10plus;
+  }
+
+  if (pt >= ptSplit && std::abs(eta) >= ebeeSplit) {
+    iCategory = CAT_EE_PT10plus;
+  }
+
+  return iCategory;
+}
+
+bool ElectronMVAEstimatorRun2Fall17::
+isEndcapCategory(int category ) const {
+
+  bool isEndcap = false;
+  if( category == CAT_EE_PT5to10 || category == CAT_EE_PT10plus ) {
+    isEndcap = true;
+  }
+
+  return isEndcap;
+}
+
+// A function that should work on both pat and reco objects
+std::vector<float> ElectronMVAEstimatorRun2Fall17::
+fillMVAVariables(const edm::Ptr<reco::Candidate>& particle, const edm::Event& iEvent) const {
+
+  //
+  // Declare all value maps corresponding to the products we defined earlier
+  //
+  edm::Handle<double> theRho;
```

Comment:
Why to repeat this at every particle?
Please get all collections only once per event